File: go/bob/bob.go
package bob

import "strings"
import "regexp"

func Hey(remark string) string {
	remark_ := strings.TrimSpace(remark)
	question := isQuestion(remark_)
	upper := isUpper(remark_)

	if isBlank(remark_) {
		return "Fine. Be that way!"
	} else if question && upper {
		return "Calm down, I know what I'm doing!"
	} else if question {
		return "Sure."
	} else if upper {
		return "Whoa, chill out!"
	} else {
		return "Whatever."
	}
}

func isBlank(str string) bool {
	return str == ""
}

func isUpper(str string) bool {
	is_lower := strings.ToLower(str) == str
	return !is_lower && strings.ToUpper(str) == str
}

func isQuestion(str string) bool {
	question_regex := regexp.MustCompile("\\?$")
	return question_regex.MatchString(str)
}

File: haskell/bob/src/Bob.hs
module Bob
  ( responseFor
  ) where

import Data.Char (isSpace, toLower, toUpper)
import Data.List (dropWhileEnd)

responseFor :: String -> String
responseFor input =
  let trimmedInput = trim input
      upperInput = map toUpper trimmedInput
      lowerInput = map toLower trimmedInput
      someLetters = upperInput /= lowerInput
      isSilence = trimmedInput == ""
      isShout = someLetters && upperInput == trimmedInput
      isQuestion = last trimmedInput == '?'
   in if isSilence
        then "Fine. Be that way!"
        else if isShout && isQuestion
               then "Calm down, I know what I'm doing!"
               else if isQuestion
                      then "Sure."
                      else if isShout
                             then "Whoa, chill out!"
                             else "Whatever."

trim :: String -> String
trim = dropWhileEnd isSpace . dropWhile isSpace
